v0/destinations/pinterestConversion/utils.js

```diff
@@ -0,0 +1,243 @@
+/* eslint-disable no-param-reassign */
+const sha256 = require("sha256");
+const { EventType } = require("../../../constants");
+const {
+  CustomError,
+  constructPayload,
+  isDefinedAndNotNull,
+  getHashFromArray
+} = require("../../util");
+const { COMMON_CONFIGS } = require("./config");
+
+const VALID_ACTION_SOURCES = ["app_android", "app_ios", "web", "offline"];
+
+const ecomEventMaps = [
+  {
+    src: ["order completed"],
+    dest: "Checkout"
+  },
+  {
+    src: ["product added"],
+    dest: "AddToCart"
+  },
+  {
+    src: ["products searched", "product list filtered"],
+    dest: "Search"
+  }
+];
+
+/**
+ *
+ * @param {*} userPayload Payload mapped from user fields
+ * @returns
+ * Further Processing the user fields following the instructions of Pinterest Conversion API
+ * Ref: https://s.pinimg.com/ct/docs/conversions_api/dist/v3.html
+ */
+const processUserPayload = userPayload => {
+  let formatValue;
+  Object.keys(userPayload).forEach(key => {
+    switch (key) {
+      case "em":
+        formatValue = userPayload[key].toString().toLowerCase();
+        userPayload[key] = [sha256(formatValue)];
+        break;
+      case "ph":
+      case "zp":
+        // zip fields should only contain digits
+        formatValue = userPayload[key].toString().replace(/[^0-9]/g, "");
+        if (key === "ph") {
+          // phone numbers should not contain leading zeros
+          formatValue = formatValue.replace(/^0+/, "");
+        }
+        userPayload[key] = [sha256(formatValue)];
+        break;
+      case "ge":
+      case "ct":
+      case "st":
+      case "country":
+        userPayload[key] = [sha256(formatValue)];
+        break;
+      case "db":
+      case "ln":
+      case "fn":
+      case "hashed_maids":
+        userPayload[key] = [userPayload[key]];
+        break;
+      default:
+        userPayload[key] = String(userPayload[key]);
+    }
+  });
+  return userPayload;
+};
+
+/**
+ *
+ * @param {*} message
+ * @returns
+ * Maps the required common parameters accross event types. Checks for the correct
+ * action source types and deduces opt_out status
+ * Ref: https://s.pinimg.com/ct/docs/conversions_api/dist/v3.html
+ */
+const processCommonPayload = message => {
+  const commonPayload = constructPayload(message, COMMON_CONFIGS);
+  const presentActionSource = commonPayload.action_source;
+  if (
+    presentActionSource &&
+    !VALID_ACTION_SOURCES.includes(presentActionSource.toLowerCase())
+  ) {
+    throw new CustomError(
+      "Action source must be one of 'app_android', 'app_ios', 'web', 'offline' ",
+      400
+    );
+  }
+  const adTrackingEnabled = message.context?.device?.adTrackingEnabled;
+  
+  if(isDefinedAndNotNull(adTrackingEnabled)) {
+        if(adTrackingEnabled === true) {
```

Comment:
We are not considering the case where `adTrackingEnabled` is a string and the value is "true"

Reply:
Because in other test cases It has been noticed as a boolean, not a string. Should we change that?

Reply:
Okay, then no need to provide support for `string`.